File: knossos-rs/src/metis.rs
//! Metis: planning.
//!
//! The plan is a separate artifact from its execution. That is the whole point
//! of splitting Metis from Talos — a plan you can print, log and disagree with
//! before any file is touched is worth more than an intention buried in a
//! model's first turn.
//!
//! The plan is requested through a tool call rather than parsed out of prose,
//! so its structure is enforced by the engine's schema validation. Local
//! models being what they are, a prose fallback exists — but it is a fallback,
//! not the design.

use anyhow::Result;
use serde_json::json;

use crate::engine::{self, Engine, Message, Request, ToolDef};
use crate::scribe::SymbolIndex;
use crate::themis::{Themis, PLANNER_ROLE};

/// A plan longer than this is a symptom — either the task wants splitting, or
/// the model is narrating rather than planning. Truncated rather than refused,
/// because a too-long plan is still more use than none.
pub const MAX_STEPS: usize = 8;

/// Tasks below this word count rarely gain from a plan, and planning them
/// costs an engine turn that execution could have had.
const TRIVIAL_WORDS: usize = 6;

/// Whether a task is big enough that a plan earns its turn.
///
/// Deliberately crude. The cost of planning a trivial task is one wasted
/// turn; the cost of *not* planning a large one is a flat loop with no
/// structure. So this errs toward planning, and only skips what is
/// obviously atomic.
pub fn worth_planning(task: &str) -> bool {
    task.split_whitespace().count() >= TRIVIAL_WORDS
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize)]
pub struct Plan {
    pub steps: Vec<String>,
}

impl Plan {
    pub fn is_empty(&self) -> bool {
        self.steps.is_empty()
    }

    /// Render for injection into Talos's opening message.
    pub fn render(&self) -> String {
        self.steps
            .iter()
            .enumerate()
            .map(|(i, s)| format!("{}. {}", i + 1, s))
            .collect::<Vec<_>>()
            .join("\n")
    }
}

fn plan_tool() -> ToolDef {
    ToolDef {
        name: "submit_plan".to_string(),
        description: format!(
            "Submit the ordered steps for this task. Between 1 and {MAX_STEPS} steps, each a \
             concrete action on the codebase."
        ),
        input_schema: json!({
            "type": "object",
            "properties": {
                "steps": {
                    "type": "array",
                    "items": {"type": "string"},
                    "description": "Ordered, concrete steps"
                }
            },
            "required": ["steps"]
        }),
    }
}

pub async fn plan(
    eng: &dyn Engine,
    themis: &Themis,
    scribe: &SymbolIndex,
    task: &str,
    max_tokens: u32,
) -> Result<Plan> {
    let system = themis.system_prompt(PLANNER_ROLE, Some(scribe));
    let req = Request::new(
        system,
        vec![Message::user_text(format!(
            "Task: {task}\n\nProduce a short plan by calling submit_plan. Keep it to the \
             smallest number of steps that actually completes the task."
        ))],
    )
    .with_tools(vec![plan_tool()])
    .with_max_tokens(max_tokens);

    let resp = match engine::complete(eng, &req).await {
        Ok(resp) => resp,
        // A planner must not end a run. The task as its own step keeps every
        // caller on one path — Talos always takes a plan.
        Err(_) => return Ok(degenerate(task)),
    };

    for (_, name, input) in resp.tool_uses() {
        if name != "submit_plan" {
            continue;
        }
        let steps = tidy(&steps_from_value(input));
        if !steps.is_empty() {
            return Ok(Plan { steps });
        }
    }

    let text = resp.text();
    let steps = tidy(&recover_truncated_json(&text));
    if !steps.is_empty() {
        return Ok(Plan { steps });
    }

    // Fallback: the engine described a plan instead of submitting one.
    let steps = tidy(&parse_prose_plan(&text));
    if steps.is_empty() {
        return Ok(degenerate(task));
    }
    Ok(Plan { steps })
}

fn degenerate(task: &str) -> Plan {
    let collapsed = task.split_whitespace().collect::<Vec<_>>().join(" ");
    Plan {
        steps: vec![if collapsed.is_empty() {
            task.to_string()
        } else {
            collapsed
        }],
    }
}

fn steps_from_value(input: &serde_json::Value) -> Vec<String> {
    let raw = input.get("steps").or_else(|| {
        input
            .get("args")
            .or_else(|| input.get("input"))
            .and_then(|inner| inner.get("steps"))
    });
    match raw {
        Some(serde_json::Value::Array(items)) => items
            .iter()
            .filter_map(|s| s.as_str())
            .map(|s| s.to_string())
            .collect(),
        _ => Vec::new(),
    }
}

/// Strip, drop blanks and duplicates, and cap the length.
fn tidy(steps: &[String]) -> Vec<String> {
    let mut seen = std::collections::BTreeSet::new();
    let mut out = Vec::new();
    for step in steps {
        let text = step.split_whitespace().collect::<Vec<_>>().join(" ");
        if text.is_empty() {
            continue;
        }
        let key = text.to_ascii_lowercase();
        if !seen.insert(key) {
            continue;
        }
        out.push(text);
        if out.len() >= MAX_STEPS {
            break;
        }
    }
    out
}

/// Recover a plan whose JSON lost its closing brackets.
///
/// Small models drop the last `}` remarkably often. Guessing at a
/// half-parsed *action* is how the wrong file gets written; a plan is
/// inert text the executor may ignore, so the same caution costs a whole
/// planning turn and buys nothing.
///
/// Narrow on purpose. Only unclosed brackets are appended, in the order
/// that closes them; nothing in the content is altered, and a block that
/// still will not parse is abandoned.
fn recover_truncated_json(reply: &str) -> Vec<String> {
    for body in fence_bodies(reply) {
        if !body.contains("submit_plan") {
            continue;
        }
        let Some(closers) = unclosed_brackets(body) else {
            continue;
        };
        if closers.is_empty() {
            // Complete JSON in a fence: still try to parse it, because a
            // model that wrote a valid `submit_plan` block as text (no
            // native tool call) should not have to fall through to prose.
            if let Some(steps) = parse_plan_json(body) {
                return steps;
            }
            continue;
        }
        let mut repaired = body.to_string();
        repaired.extend(closers.into_iter().rev());
        if let Some(steps) = parse_plan_json(&repaired) {
            return steps;
        }
    }
    Vec::new()
}

fn parse_plan_json(body: &str) -> Option<Vec<String>> {
    let payload: serde_json::Value = serde_json::from_str(body).ok()?;
    let steps = steps_from_value(&payload);
    if steps.is_empty() {
        None
    } else {
        Some(steps)
    }
}

/// Bodies of fenced blocks, including an unterminated fence — a truncated
/// reply may never close it.
fn fence_bodies(reply: &str) -> Vec<&str> {
    let mut out = Vec::new();
    let mut rest = reply;
    while let Some(start) = rest.find("```") {
        let after_ticks = &rest[start + 3..];
        let after_lang = match after_ticks.find('\n') {
            Some(i) => &after_ticks[i + 1..],
            None => break,
        };
        match after_lang.find("```") {
            Some(end) => {
                out.push(&after_lang[..end]);
                rest = &after_lang[end + 3..];
            }
            None => {
                out.push(after_lang);
                break;
            }
        }
    }
    out
}

/// Closers needed to balance `{`/`[`, or `None` if a string is unclosed
/// (we cannot decide) or the brackets are already balanced / over-closed.
fn unclosed_brackets(body: &str) -> Option<Vec<char>> {
    let mut stack = Vec::new();
    let mut in_string = false;
    let mut escaped = false;
    for c in body.chars() {
        if in_string {
            if escaped {
                escaped = false;
            } else if c == '\\' {
                escaped = true;
            } else if c == '"' {
                in_string = false;
            }
            continue;
        }
        match c {
            '"' => in_string = true,
            '{' => stack.push('}'),
            '[' => stack.push(']'),
            '}' | ']' => match stack.pop() {
                Some(expected) if expected == c => {}
                _ => return None,
            },
            _ => {}
        }
    }
    if in_string {
        return None;
    }
    Some(stack)
}

/// Pull numbered or bulleted lines out of prose.
fn parse_prose_plan(text: &str) -> Vec<String> {
    let mut out = Vec::new();
    for line in text.lines() {
        let t = line.trim();
        if t.is_empty() {
            continue;
        }

        let stripped = t
            .strip_prefix("- ")
            .or_else(|| t.strip_prefix("* "))
            .map(str::to_string)
            .or_else(|| {
                // "1. step" / "2) step"
                let mut chars = t.char_indices();
                let digits: String = chars
                    .by_ref()
                    .take_while(|(_, c)| c.is_ascii_digit())
                    .map(|(_, c)| c)
                    .collect();
                if digits.is_empty() {
                    return None;
                }
                let rest = t[digits.len()..].trim_start();
                rest.strip_prefix('.')
                    .or_else(|| rest.strip_prefix(')'))
                    .map(|r| r.trim().to_string())
            });

        if let Some(s) = stripped {
            if !s.is_empty() {
                out.push(s);
            }
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::engine::mock::{text_response, MockEngine};
    use crate::engine::{Content, Response, StopReason, Usage};

    fn scribe() -> SymbolIndex {
        let dir = tempfile::tempdir().unwrap();
        SymbolIndex::build(dir.path()).unwrap()
    }

    #[tokio::test]
    async fn uses_the_submitted_plan_when_the_tool_is_called() {
        let eng = MockEngine::new(vec![Response {
            content: vec![Content::ToolUse {
                id: "1".into(),
                name: "submit_plan".into(),
                input: json!({"steps": ["read lib.rs", "add the flag", "run cargo check"]}),
            }],
            stop_reason: StopReason::ToolUse,
            usage: Usage::default(),
        }]);

        let p = plan(&eng, &Themis::from_text("x"), &scribe(), "add a flag", 1024)
            .await
            .unwrap();
        assert_eq!(p.steps.len(), 3);
        assert_eq!(p.steps[0], "read lib.rs");
        assert!(p.render().starts_with("1. read lib.rs"));
    }

    #[tokio::test]
    async fn falls_back_to_parsing_prose() {
        let eng = MockEngine::new(vec![text_response(
            "Here is my plan:\n1. Read the file\n2. Make the edit\n- Verify with cargo",
        )]);
        let p = plan(&eng, &Themis::from_text("x"), &scribe(), "task", 1024)
            .await
            .unwrap();
        assert_eq!(
            p.steps,
            vec!["Read the file", "Make the edit", "Verify with cargo"]
        );
    }

    #[tokio::test]
    async fn an_unparseable_reply_still_yields_a_usable_plan() {
        let eng = MockEngine::new(vec![text_response("I'm not sure what to do.")]);
        let p = plan(&eng, &Themis::from_text("x"), &scribe(), "add a flag", 1024)
            .await
            .unwrap();
        assert_eq!(p.steps, vec!["add a flag"]);
        assert!(!p.is_empty());
    }

    #[tokio::test]
    async fn the_planner_prompt_carries_the_constitution() {
        let eng = MockEngine::new(vec![text_response("- do it")]);
        plan(
            &eng,
            &Themis::from_text("CONSTITUTION MARKER"),
            &scribe(),
            "t",
            512,
        )
        .await
        .unwrap();
        let seen = eng.seen.lock().unwrap();
        assert!(seen[0].system.contains("CONSTITUTION MARKER"));
        assert!(seen[0].system.contains("Metis"));
    }

    #[test]
    fn prose_parser_handles_both_numbering_styles() {
        let s = parse_prose_plan("1. alpha\n2) beta\n- gamma\n* delta\nnot a step");
        assert_eq!(s, vec!["alpha", "beta", "gamma", "delta"]);
    }

    #[test]
    fn a_plan_missing_its_closing_brace_is_recovered() {
        let reply = "```json\n{\"tool\": \"submit_plan\", \"args\": {\"steps\": [\"read it\", \"fix it\"]}\n```";
        assert_eq!(recover_truncated_json(reply), ["read it", "fix it"]);
    }

    #[test]
    fn a_plan_truncated_mid_array_is_recovered() {
        let reply = "```json\n{\"tool\": \"submit_plan\", \"args\": {\"steps\": [\"read it\"";
        assert_eq!(recover_truncated_json(reply), ["read it"]);
    }

    #[test]
    fn a_brace_inside_a_string_is_not_counted() {
        let reply = "```json\n{\"tool\": \"submit_plan\", \"args\": {\"steps\": [\"handle the { case\"]}\n```";
        assert_eq!(recover_truncated_json(reply), ["handle the { case"]);
    }

    #[test]
    fn genuinely_broken_json_is_not_guessed_at() {
        let reply = "```json\n{\"tool\": \"submit_plan\", \"args\": {\"steps\": [oh dear]}}\n```";
        assert!(recover_truncated_json(reply).is_empty());
    }

    #[test]
    fn a_truncated_block_for_another_tool_is_ignored() {
        let reply = "```json\n{\"tool\": \"write_file\", \"args\": {\"path\": \"a.py\"";
        assert!(recover_truncated_json(reply).is_empty());
    }

    #[test]
    fn blank_and_duplicate_steps_are_dropped() {
        let steps = tidy(&[
            "read it".into(),
            "  ".into(),
            "read it".into(),
            "READ IT".into(),
            "fix it".into(),
        ]);
        assert_eq!(steps, ["read it", "fix it"]);
    }

    #[test]
    fn a_runaway_plan_is_capped_not_refused() {
        let steps: Vec<String> = (0..50).map(|i| format!("step {i}")).collect();
        assert_eq!(tidy(&steps).len(), MAX_STEPS);
    }

    #[test]
    fn trivial_tasks_skip_the_planning_turn() {
        assert!(!worth_planning("fix it"));
        assert!(!worth_planning("add a flag"));
        assert!(worth_planning("add a flag to the config loader"));
    }

    #[tokio::test]
    async fn a_broken_engine_does_not_raise() {
        let eng = MockEngine::new(vec![]); // no scripted replies: complete fails
        let p = plan(
            &eng,
            &Themis::from_text("x"),
            &scribe(),
            "a task long enough to plan",
            512,
        )
        .await
        .unwrap();
        assert_eq!(p.steps, ["a task long enough to plan"]);
    }
}

File: model/knossos/metis.py
"""Metis: planning, as an artifact separate from its execution.

That separation is the whole point of splitting this from Talos. A plan you can
read, log, and disagree with *before* any file is touched is worth more than an
intention buried in the model's first turn -- and once it exists as a list, the
editor can render it as a checklist and the executor can be held to it.

The plan is requested as a tool call rather than parsed out of prose, so its
shape is constrained by a schema the engine already knows how to emit. Local
models being what they are, a prose fallback exists; it is a fallback, not the
design. And if both fail, the task itself becomes a one-step plan, because a
planner that can return *nothing* would make planning a new way for a run to
fail before it starts.

Planning costs one engine turn. `Ariadne`'s budget is for execution, so a task
small enough not to need a plan should skip it -- see `worth_planning`.

Ported from `knossos-rs/src/metis.rs`, which is the reference implementation.
"""
from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from typing import Callable, List, Optional, Sequence

from .jsonrpc import log
from .tools import ToolRegistry, parse_calls

__all__ = ["Plan", "Metis", "PLANNER_ROLE", "MAX_STEPS", "worth_planning"]

#: A plan longer than this is a symptom -- either the task wants splitting, or
#: the model is narrating rather than planning. Truncated rather than refused,
#: because a too-long plan is still more use than none.
MAX_STEPS = 8

PLANNER_ROLE = """\
You are Metis, the planner. You do not edit files and you do not run commands.
You produce the shortest ordered list of concrete actions that completes the
task, and nothing else.

Rules:
- Each step is one concrete action on this codebase, phrased so someone else
  could carry it out without asking you what you meant.
- Prefer fewer steps. A one-step plan is correct for a one-step task.
- Do not include "verify" or "run the tests" as a final step: verification runs
  automatically after execution and is not yours to schedule.
- Do not restate the task as step 1.
"""

PLAN_TOOL = {
    "type": "object",
    "properties": {
        "steps": {"type": "array", "items": {"type": "string"},
                  "description": "Ordered, concrete steps"},
    },
    "required": ["steps"],
}

#: Bullets and both numbering styles: "- x", "* x", "1. x", "2) x".
_BULLET = re.compile(r"^\s*(?:[-*]|\d+[.)])\s+(.*\S)\s*$")

#: A fenced block's body. Matches an *unterminated* fence too, which is the
#: whole point -- a truncated reply may never close it.
_FENCE_BODY = re.compile(r"```(?:json)?\s*\n(.*?)(?:```|\Z)", re.DOTALL)

#: Tasks below this word count rarely gain from a plan, and planning them costs
#: an engine turn that execution could have had.
_TRIVIAL_WORDS = 6


def worth_planning(task: str) -> bool:
    """Whether a task is big enough that a plan earns its turn.

    Deliberately crude. The cost of planning a trivial task is one wasted turn;
    the cost of *not* planning a large one is a flat loop with no structure. So
    this errs toward planning, and only skips what is obviously atomic.
    """
    words = task.split()
    if len(words) < _TRIVIAL_WORDS:
        return False
    # A task naming several things, or several sentences, wants a plan.
    return True


@dataclass
class Plan:
    """An ordered list of concrete steps. Possibly empty; never None."""

    steps: List[str] = field(default_factory=list)
    #: True when no plan could be extracted and the task became its own step.
    degenerate: bool = False

    def __bool__(self) -> bool:
        return bool(self.steps)

    def __len__(self) -> int:
        return len(self.steps)

    def render(self) -> str:
        return "\n".join(f"{i + 1}. {s}" for i, s in enumerate(self.steps))


class Metis:
    """Turns a task into a plan, using one engine turn.

    Holds no state between calls: a plan is a function of the task and the
    context it was given, and caching one across tasks is how a stale plan gets
    executed against the wrong request.
    """

    def __init__(self, engine, tools: Optional[ToolRegistry] = None,
                 constitution: str = "", max_steps: int = MAX_STEPS) -> None:
        self.engine = engine
        self.tools = tools
        self.constitution = constitution
        self.max_steps = max_steps

    # ------------------------------------------------------------------ api

    def plan(self, task: str, context: str = "",
             cancelled: Optional[Callable[[], bool]] = None) -> Plan:
        """Produce a plan for `task`. Always returns a usable one."""
        alive = cancelled or (lambda: False)
        try:
            reply = "".join(self._stream(self._prompt(task), context, alive))
        except Exception as exc:                     # a planner must not end a run
            log(f"[metis] planning failed, falling back to the task itself: {exc}")
            return self._degenerate(task)

        if alive():
            return self._degenerate(task)

        steps = (self._from_tool_call(reply)
                 or self._from_truncated_json(reply)
                 or self._from_prose(reply))
        if not steps:
            log("[metis] no plan could be extracted from the reply")
            return self._degenerate(task)
        return Plan(steps=self._tidy(steps))

    # -------------------------------------------------------------- prompt

    def _prompt(self, task: str) -> str:
        parts = [PLANNER_ROLE]
        if self.constitution:
            parts.append(f"\n# Constitution\n\nThese apply to everything you do.\n\n"
                         f"{self.constitution}")
        if self.tools is not None:
            # The planner does not call these, but a plan written without
            # knowing what the executor can do is a plan it cannot carry out.
            parts.append("\n# The executor has these tools\n\n"
                         + ", ".join(self.tools.names))
        parts.append(
            "\n# Submitting the plan\n\n"
            "Emit exactly one fenced JSON block and no other JSON:\n\n"
            '```json\n{"tool": "submit_plan", "args": {"steps": ["...", "..."]}}\n```\n'
            f"\nBetween 1 and {self.max_steps} steps.")
        parts.append(f"\n# Task\n\n{task}")
        return "\n".join(parts)

    def _advertise(self) -> None:
        """Offer `submit_plan` as a real tool, not only as prose.

        A model that prefers native tool calls will make one whatever the prompt
        says -- and if the only tool it has been *given* is nothing, it invents
        a shape and the plan is unparseable. Observed with gemma4:e4b, which
        called tools happily all through execution and produced no usable plan
        at all, so stepwise execution never engaged and the whole planning path
        silently did nothing.

        Only `submit_plan` is offered. Handing the planner the executor's tools
        would invite it to start editing files during the turn whose entire
        purpose is to decide what to do before anything is touched.
        """
        if not hasattr(self.engine, "tool_schema"):
            return
        self.engine.tool_schema = [{
            "type": "function",
            "function": {
                "name": "submit_plan",
                "description": (f"Submit the ordered steps for this task. Between "
                                f"1 and {self.max_steps} steps, each a concrete "
                                f"action on the codebase."),
                "parameters": PLAN_TOOL,
            },
        }]

    def _stream(self, prompt: str, context: str, alive):
        self._advertise()
        try:
            for chunk in self.engine.generate(prompt, context, alive):
                if type(chunk).__name__ == "Thought":
                    continue                         # reasoning is not the plan
                yield str(chunk)
        finally:
            # Talos re-advertises its own tools before executing, but a planner
            # that leaves `submit_plan` on the engine would let the executor
            # call it -- so clear rather than rely on that ordering.
            if hasattr(self.engine, "tool_schema"):
                self.engine.tool_schema = None

    # ------------------------------------------------------------ extraction

    @staticmethod
    def _from_tool_call(reply: str) -> List[str]:
        """The designed path: a `submit_plan` call, native or fenced.

        The engine normalises native `tool_calls` into fenced blocks, so both
        wire formats arrive here identically.
        """
        _prose, calls = parse_calls(reply)
        for call in calls:
            if call.name != "submit_plan":
                continue
            steps = call.args.get("steps")
            if isinstance(steps, list):
                return [str(s) for s in steps]
        return []

    @staticmethod
    def _from_truncated_json(reply: str) -> List[str]:
        """Recover a plan whose JSON lost its closing brackets.

        Small models drop the last `}` remarkably often. `parse_calls` refuses
        such a block, and for a *tool call* that is exactly right -- guessing at
        a half-parsed action is how the wrong file gets written. A plan is not
        an action: it is inert text that the executor may ignore, so the same
        caution costs a whole planning turn and buys nothing.

        Narrow on purpose. Only unclosed brackets are appended, in the order
        that closes them; nothing in the content is altered, and a block that
        still will not parse is abandoned. Observed with gemma4:e4b, which
        emitted a complete and correct plan missing one closing brace.
        """
        for match in _FENCE_BODY.finditer(reply):
            body = match.group(1).strip()
            if "submit_plan" not in body:
                continue
            stack: List[str] = []
            in_string = escaped = False
            for char in body:
                if in_string:
                    if escaped:
                        escaped = False
                    elif char == "\\":
                        escaped = True
                    elif char == '"':
                        in_string = False
                    continue
                if char == '"':
                    in_string = True
                elif char in "{[":
                    stack.append("}" if char == "{" else "]")
                elif char in "}]" and stack:
                    stack.pop()
            if in_string or not stack:
                continue                             # not a truncation we can fix
            try:
                payload = json.loads(body + "".join(reversed(stack)))
            except json.JSONDecodeError:
                continue
            steps = (payload.get("args") or {}).get("steps")
            if isinstance(steps, list) and steps:
                log("[metis] recovered a plan from truncated JSON")
                return [str(s) for s in steps]
        return []

    @staticmethod
    def _from_prose(reply: str) -> List[str]:
        """The fallback: numbered or bulleted lines.

        Only lines that are *marked* as list items count. Treating every line as
        a step turns a paragraph of hedging into an eight-step plan.
        """
        out: List[str] = []
        for line in reply.splitlines():
            match = _BULLET.match(line)
            if match:
                out.append(match.group(1))
        return out

    def _tidy(self, steps: Sequence[str]) -> List[str]:
        """Strip, drop blanks and duplicates, and cap the length."""
        seen, out = set(), []
        for step in steps:
            text = " ".join(str(step).split())
            if not text or text.lower() in seen:
                continue
            seen.add(text.lower())
            out.append(text)
            if len(out) >= self.max_steps:
                break
        return out

    @staticmethod
    def _degenerate(task: str) -> Plan:
        """The task as its own single step.

        A plan that is only the task restates nothing useful, but it keeps every
        downstream caller on one code path -- `Talos.run` takes a plan, the ACP
        layer renders one, and neither has to special-case an absent plan.
        """
        return Plan(steps=[" ".join(task.split())], degenerate=True)
